src/client/dfuse/ops/mkdir.c

```diff
@@ -24,41 +24,65 @@
 #include "dfuse_common.h"
 #include "dfuse.h"
 
-#define REQ_NAME request
-#define POOL_NAME mkdir_da
-#define TYPE_NAME entry_req
-#include "dfuse_ops.h"
-
-static const struct dfuse_request_api api = {
-	.on_result	= dfuse_entry_cb,
-};
-
 void
 dfuse_cb_mkdir(fuse_req_t req, fuse_ino_t parent, const char *name, mode_t mode)
 {
 	struct dfuse_projection_info	*fs_handle = fuse_req_userdata(req);
-	struct entry_req		*desc = NULL;
+	struct dfuse_inode_entry	*inode = NULL;
+	struct dfuse_inode_entry	*parent_inode;
+	d_list_t			*rlink;
 	int rc;
 
 	DFUSE_TRA_INFO(fs_handle, "Parent:%lu '%s'", parent, name);
-	DFUSE_REQ_INIT_REQ(desc, fs_handle, api, req, rc);
-	if (rc)
-		D_GOTO(err, rc);
 
-	strncpy(desc->ie->name, name, NAME_MAX);
-	desc->ie->parent = parent;
-	desc->da = fs_handle->mkdir_da;
+	rlink = d_hash_rec_find(&fs_handle->inode_ht, &parent, sizeof(parent));
+	if (!rlink) {
+		DFUSE_TRA_ERROR(fs_handle, "Failed to find inode %lu",
+				parent);
+		D_GOTO(err, rc = ENOENT);
+	}
+
+	parent_inode = container_of(rlink, struct dfuse_inode_entry, ie_htl);
+
+	D_ALLOC_PTR(inode);
+	if (!inode) {
+		D_GOTO(err, rc = ENOMEM);
+	}
+
+	DFUSE_TRA_INFO(parent_inode, "parent");
+
+	/* mkdir with the correct parent */
+	rc = dfs_mkdir(fs_handle->fsh_dfs, parent_inode->obj, name, mode);
+	if (rc != -DER_SUCCESS) {
```

Comment:
i have in dfuse_hl something like a daos error code to errno code convertor. not sure if you are doing something similar, but if not you can reuse that.

Reply:
Yes, this is a problem and was never easy to get right in IOF.  Currently DFUSE_REPLY_ERR_RAW converts negative numbers to EIO, but this is a blunt hammer.

The error_convert() function in dfuse_hl.c seems to return a mixture of CaRT and system error numbers, and to compound the issue always negative values.

I think FUSE high level uses negative error numbers as there cases where positive numbers are used for size and negative numbers for errors, however FUSE low level avoids this and correctly uses positive error numbers in all cases.